### PR TITLE
Resources improvements for native macOS app

### PR DESCRIPTION
**Pull request by Thomas-Vos**
First commit fixes compose resources for native macOS app (see currently broken chat app using `runDebugExecutableMacosArm64`).

Second commit adds support for embedding resources into native macOS framework. Similar to how it works on iOS. This allows using the macOS compose code in an Xcode project. Just like iOS, using kotlin code as framwork. Then call the main function from the Xcode project to run the compose app.

## Testing
First commit: run chat app using `runDebugExecutableMacosArm64`
Second commit: Tested by adding macOS support to the iOS Xcode project in chat app (not in this PR). Can add the sample app if needed.

### CODE REVIEW

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/resources/IosResources.kt [hunk 1]
```diff
@@ -127,7 +127,8 @@ private fun Framework.getFinalResourcesDir(): Provider<Directory> {
             .zip(
                 providers.environmentVariable("CONTENTS_FOLDER_PATH")
             ) { builtProductsDir, contentsFolderPath ->
-                File("$builtProductsDir/$contentsFolderPath/$IOS_COMPOSE_RESOURCES_ROOT_DIR").canonicalPath
+                val pathSuffix = if (target.isMacTarget()) "/Resources" else ""
+                File("$builtProductsDir/$contentsFolderPath$pathSuffix/$IOS_COMPOSE_RESOURCES_ROOT_DIR").canonicalPath
```
Comment (terrakok) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/resources/IosResources.kt:131]: I don't like the idea to hardcode a mac specific path here.
It is better to make new function for mac targets and support cocoapods as well then.

**Review by terrakok — CHANGES_REQUESTED** [review]
Tests for the new logic are required

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/resources/IosResources.kt (hunk 1 above)
Comment (Thomas-Vos) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/resources/IosResources.kt]: Fixed the path by changing `CONTENTS_FOLDER_PATH` to `UNLOCALIZED_RESOURCES_FOLDER_PATH`. Now it is the same for iOS and macOS.

Comment (Thomas-Vos) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/resources/IosResources.kt]: Also added a test for cocoapods (copied from iOS with minor changes), so I think all should work fine.

**Comment by Thomas-Vos** [comment]
@terrakok thanks a lot for your review! Thanks for helping me with reviewing some fixes for macOS issues so I can start using it in my app.

I added tests to the PR as you asked. However, the tests are failing on CI because they use an outdated snapshot of the resources library. For the tests to pass, they must use the latest version of the resources library. This is because there are changes in the resources library in order to support macOS. When forcing a local build of the library, it passes on my own mac:

Screenshot 2024-11-30 at 18 35 26


Is there any way to fix this? Or is this expected to happen until the next compose snapshot is released with the changes of this PR?

File: components/resources/library/src/macosMain/kotlin/org/jetbrains/compose/resources/ResourceReader.macos.kt [hunk 2]
```diff
@@ -39,12 +39,15 @@ internal actual fun getPlatformResourceReader(): ResourceReader = object : Resou
     private fun getPathOnDisk(path: String): String {
         val fm = NSFileManager.defaultManager()
         val currentDirectoryPath = fm.currentDirectoryPath
+        val pathFix = path.removePrefix("composeResources/").substringAfter("/")
```
Comment (terrakok) [inline components/resources/library/src/macosMain/kotlin/org/jetbrains/compose/resources/ResourceReader.macos.kt:42]: I find this line too dangerous. It mixes `removePrefix` and `substringAfter` that is hard to predict a final result:
`"composeResources/foo/bar/baz" -> "bar/baz"`
`"/foo/bar/baz" -> "foo/bar/baz"`
`"/composeResources/foo/bar/baz" -> "composeResources/foo/bar/baz"`

Could you please refactor it to use Path API and add a comment here what and why was fixed, please.

**Review by terrakok — CHANGES_REQUESTED** [review]

**Review by terrakok — APPROVED** [review]

**Comment by terrakok** [comment]
I reverted the commit because 
1) tests are failing
```
> Task :appModule:macosArm64Test FAILED

ComposeAppTest.checkApp[macosArm64] FAILED
    org.jetbrains.compose.resources.MissingResourceException at /Users/Konstantin.Tskhovrebov/Desktop/junit17706467430052708708/appModule/src/commonTest/kotlin/ComposeAppTest.kt:25

1 test completed, 1 failed
[Incubating] Problems report is available at: file:///Users/Konstantin.Tskhovrebov/Desktop/junit17706467430052708708/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.
```

2) There is a problem with Skiko publication https://youtrack.jetbrains.com/issue/CMP-3859

**Comment by Thomas-Vos** [comment]
This PR was reverted, see new PR: https://github.com/JetBrains/compose-multiplatform/pull/5219

**Comment by terrakok** [comment]
I reverted the change again because `org.jetbrains.compose.test.tests.integration.ResourcesTest.testMultiModuleResources(kotlin=2.1.0, gradle=8.8, agp=8.5.0)` fails on Linux with:
```
Caused by: org.gradle.internal.resolve.ModuleVersionNotFoundException: Could not find me.sample.library:cmplib-macosarm64:1.0.
Searched in the following locations:
  - https://repo.maven.apache.org/maven2/me/sample/library/cmplib-macosarm64/1.0/cmplib-macosarm64-1.0.pom
  - https://dl.google.com/dl/android/maven2/me/sample/library/cmplib-macosarm64/1.0/cmplib-macosarm64-1.0.pom
  - https://packages.jetbrains.team/files/p/mpp/kgp-for-compose2/me/sample/library/cmplib-macosarm64/1.0/cmplib-macosarm64-1.0.pom
  - https://maven.pkg.jetbrains.space/public/p/compose/dev/me/sample/library/cmplib-macosarm64/1.0/cmplib-macosarm64-1.0.pom
  - https://maven.pkg.jetbrains.space/kotlin/p/kotlin/dev/me/sample/library/cmplib-macosarm64/1.0/cmplib-macosarm64-1.0.pom
  - file:/root/.m2/repository/me/sample/library/cmplib-macosarm64/1.0/cmplib-macosarm64-1.0.pom
  - file:/mnt/agent/temp/buildTmp/junit3414287715548457763/my-mvn/me/sample/library/cmplib-macosarm64/1.0/cmplib-macosarm64-1.0.pom
Required by:
    project :appModule > me.sample.library:cmplib:1.0
```

**Pull request merged**